### PR TITLE
Issue debug message for invalid types

### PR DESCRIPTION
**Pull request by AlexandreYang**
When using metric_type in  jmxfetch metrics definition, if the value is anything other than `gauge`, `histogram`, `monotonic_count` it will default to “counter” (aka rate).  https://github.com/DataDog/jmxfetch/blob/f39e75a8a3d5c83890d01f6752cea4dc0aef679c/src/main/java/org/datadog/jmxfetch/reporter/Reporter.java#L122

This seems not ideal since multiple values for `metric_type` can lead this same `counter/rate` behaviour including typos like `gauuge`.

The fix: issue warning/deprecation for anything other than `counter` and `rate`.

We keep both counter and rate as valid since they are already in use and documented here: https://docs.datadoghq.com/integrations/java/?tab=host#the-attribute-filter

### CODE REVIEW

**Comment by AlexandreYang** [comment]
Related PR: https://github.com/DataDog/integrations-core/pull/7275

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java [hunk 1]
```diff
@@ -119,7 +119,15 @@ public void sendMetrics(List<Metric> metrics, String instanceName, boolean canon
...
 
-            } else { // The metric should be 'counter'
+            } else {
+                // `counter` and `rate` are equivalent and both accepted as valid.
+                // We continue accepting other cases for backward compatibility.
+                if (!"counter".equals(metricType) && !"rate".equals(metricType)) {
+                    log.warn("Invalid metric type " + metricType + " for metric " + metricName
+                            + ". The metric will be processed as rate"
+                            + " (this is a DEPRECATED behaviour, use a valid type instead).");
+                }
```
Comment (AlexandreYang) [inline src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:129]: Note: I tried to have a logic like

```
            if ("gauge".equals(metricType) || "histogram".equals(metricType)) {
                ...
            } else if ("monotonic_count".equals(metricType)) {
                ...
            } else if ("counter".equals(metricType) || "rate".equals(metricType)) {
                sendCounter(...)
            } else {
                // issue warning and deprecation
               sendCounter(...)
            }
```

But the issue is that we will need to pass about 7 args to sendCounter, which is not great. That's why I went to this low changelog approach you see here that only add the warning.

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java [hunk 2]
```diff
@@ -119,7 +119,15 @@ public void sendMetrics(List<Metric> metrics, String instanceName, boolean canon
                 }
                 sendMetricPoint(metricType, metricName, delta, tags);
 
-            } else { // The metric should be 'counter'
+            } else {
+                // `counter` and `rate` are equivalent and both accepted as valid.
+                // We continue accepting other cases for backward compatibility.
+                if (!"counter".equals(metricType) && !"rate".equals(metricType)) {
```
Comment (truthbk) [inline src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:125]: IMO: this reads better as:
```
 if (!metricType.equals("counter") && !metricType.equals("rate")) {
    ...
}
```
But I just noticed we use this convention elsewhere in the file, so please feel free to ignore or fix both.

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java [hunk 3]
```diff
@@ -119,7 +119,15 @@ public void sendMetrics(List<Metric> metrics, String instanceName, boolean canon
                 }
                 sendMetricPoint(metricType, metricName, delta, tags);
 
-            } else { // The metric should be 'counter'
+            } else {
+                // `counter` and `rate` are equivalent and both accepted as valid.
+                // We continue accepting other cases for backward compatibility.
```
Comment (truthbk) [inline src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:124]: I would rephrase this:
```
Other types, deprecated or wrong, will default to a counter/rate type and a warning will be logged.
```

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java [hunk 4]
```diff
@@ -119,7 +119,15 @@ public void sendMetrics(List<Metric> metrics, String instanceName, boolean canon
...
+                if (!"counter".equals(metricType) && !"rate".equals(metricType)) {
+                    log.warn("Invalid metric type " + metricType + " for metric " + metricName
```
Comment (truthbk) [inline src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:126]: I think the warning level is way too verbose, if someone triggers this code path this will be all over the logs.

Comment (AlexandreYang) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:126]: That's right, the upside is that it will motivate to fix the issue.

If we use info, the quantity of output will be same and less visible.

If we use debug, people won't see it and/or ignore it.

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java (hunk 2 above)
Comment (AlexandreYang) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java]: Changed for this case and elsewhere.

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java (hunk 3 above)
Comment (AlexandreYang) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java]: Changed.

File: src/main/java/org/datadog/jmxfetch/reporter/Reporter.java (hunk 4 above)
Comment (AlexandreYang) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:127]: @truthbk An alternative is the have an boolean attribute like `this.isInvalidMetricTypeWarningLogged` to log only once:

```java
                if (!metricType.equals("counter")
                        && !metricType.equals("rate")
                        && !this.isInvalidMetricTypeWarningLogged) {
                    log.warn("Invalid metric type " + metricType + " for metric " + metricName
                            + ". The metric will be processed as rate"
                            + " (this is a DEPRECATED behaviour, use a valid type instead).");
                    this.isInvalidMetricTypeWarningLogged = true;
                }
```

WDYT?

Comment (truthbk) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java:127]: >  it will motivate to fix the issue
Yeah, that's for sure! 😁 

But some customers ingest agent logs, and those are billed, so I'm pretty sure we shouldn't log at the warn level. As to adding a flag like that, it could work. Printing it just once might not be enough, especially with the log rotation. Maybe lowering it to `debug` is enough, or if you want to make it stand out in `warn`, you could maybe use an integer and set a desired frequency to skip, and print every N iterations or something - it would add some overhead, but only for the bad types so maybe not the end of the world.

Comment (AlexandreYang) [reply src/main/java/org/datadog/jmxfetch/reporter/Reporter.java]: Let's use debug for now then :)

Changed.

**Comment by AlexandreYang** [comment]
@truthbk Could you have a look again at the PR. The comments have been addressed.

**Review by truthbk — APPROVED** [review]

**Comment by froque** [comment]
I have been bitten by the debug level. A warning level should have been used at least once.
